### ISSUE TITLE
Fix Robot Program Re-Enabling

### ISSUE THREAD

**Issue by gnawinganimal** [SouthEugeneRoboticsTeam/sertain#25]
It seems that whenever a robot is enabled then disabled while running sertain, the same instance of the program continues to run after it is enabled again.

**Comment by gnawinganimal** [comment]
I am pretty sure that we had this same problem during the past season, where we would press enable again and it would just keep doing whatever it was doing when it was disabled. And I think the problem was that we needed to cancel coroutines. @andrewda, do you remember this? Just want to make sure I am remembering correctly.

**Comment by andrewda** [comment]
Ah yeah, I remember that. I think I might have fixed that on the meanlib level, let me investigate a little.

**Comment by andrewda** [comment]
@Jamdan2 forwarded you my email thread with Tyler which has some similar stuff. A few guiding questions I have:

- Does the coroutine continue to run even when disabled? I believe that's the issue we were having with meanlib during bunnybots last year (and it seems to be what I'm describing in my emails).

- Also, where/how are you currently cancelling actions when the robot gets disabled? Meanlib has a [`Subsystem::disable()`](https://github.com/TeamMeanMachine/meanlib/blob/40a08784ba04a7df1e4f7222801282898847191f/src/main/kotlin/org/team2471/frc/lib/framework/Subsystem.kt#L98) method which cancels the subsytem's action (see implementation [here](https://github.com/TeamMeanMachine/meanlib/blob/3d0fe9936c3cafa773a69956307bb33b34e7ca02/src/main/kotlin/org/team2471/frc/lib/framework/internal/SubsystemCoordinator.kt#L174-L179)). I'm not seeing any similar logic in Sertain (though I haven't had much time to dig around), so I'm confused as to where you expect the action to be canceled.

**Comment by gnawinganimal** [comment]
Found a fix for this.
`sertain` is pretty different than meanlib, so doing the same thing as `Subsystem::disable` did not work. However, I found a solution by modifying event subscribers. I changed
```
inline fun <reified E : Event> CoroutineScope.subscribe(noinline action: suspend (E) -> Unit) =
            launch {
                events.asFlow().filter { it is E }.map { it as E }.apply { collect(action) }
            }
```
to
```
inline fun <reified E : Event> CoroutineScope.subscribe(noinline action: suspend (E) -> Unit) =
            launch {
                events.asFlow().filter { it is E }.map { it as E }.apply { collect {
                    launch {
                        action(it)
                    }
                }}
            }
```
No idea why this works, but a fix is a fix 😆 

I still need to test this for possible side effects.

**Comment by andrewda** [comment]
Strange... Glad it worked, though!

**Comment by gnawinganimal** [comment]
Solved by #26 👍 🎉 🎈
